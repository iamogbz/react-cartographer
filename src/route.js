import React from "react";
import ReactRoute from "react-router/Route";

export default class Route {
    /**
     * Create a new Route object
     * @param {{}} route shape
     */
    constructor({ name, props, renderProps, suffixes, nest } = {}) {
        this.name = name ? name.toString() : undefined;
        this.props = { ...props };
        this.renderProps = { ...renderProps };
        this.suffixes = { ...suffixes };
        this.nest = nest
            ? {
                  props: { ...nest.props },
                  renderProps: { ...nest.renderProps },
                  routes: (nest.routes || []).map(shape => new Route(shape)),
              }
            : { props: {}, renderProps: {}, routes: [] };
    }

    /**
     * Get name of the route
     * @returns {string}
     */
    get name() {
        return this._name;
    }

    /**
     * Sets name of the route
     * @param {string} value non string values are rejected and warned against
     */
    set name(value) {
        this._name = value;
        if (value && typeof value !== "string")
            console.warn(`Route name was set to non string value '${value}'`);
    }

    /**
     * Sets name of the route
     * @param {string} name
     * @returns {Route} reference to updated object
     */
    setName = name => Object.assign(this, { name });

    /**
     * Sets react router route props
     * @param {{}} props
     * @returns {Route} reference to updated object
     */
    setProps = props => Object.assign(this, { props });

    /**
     * Adds to react router route props
     * @param {{}} props
     * @returns {Route} reference to updated object
     */
    addProps = props => {
        Object.assign(this.props, props);
        return this;
    };

    rPath = path => this.addProps({ path });

    rKey = key => this.addProps({ key });

    rExact = exact => this.addProps({ exact });

    rStrict = strict => this.addProps({ strict });

    rLocation = location => this.addProps({ location });

    rSensitive = sensitive => this.addProps({ sensitive });

    rChildren = children => this.addProps({ children });

    rComponent = component => this.addProps({ component });

    rRender = render => this.addProps({ render });

    /**
     * Extra properties passed to the render of this route.
     * When this is non empty react router render property is always used
     * @param {{}} renderProps
     * @returns {Route} reference to updated object
     */
    setRenderProps = renderProps => Object.assign(this, { renderProps });

    /**
     * List of paths appended to base, result used as Route.path prop
     * @param {[]} suffixes
     * @returns {Route} reference to updated object
     */
    setSuffixes = suffixes => Object.assign(this, { suffixes });

    /**
     * Add one or more suffixes to existing route.suffixes
     * @param {{}} suffixes
     * @returns {Route} reference to updated object
     */
    addSuffixes = suffixes =>
        Object.assign(this, {
            suffixes: Object.assign(this.suffixes, suffixes),
        });

    /**
     * Deletes specified suffixes using names as keys
     * @param {...string} names
     * @returns {Route} reference to updated object
     */
    removeSuffixes = (...names) =>
        Object.assign(this, {
            suffixes: Object.keys(this.suffixes)
                .filter(key => !names.includes(key))
                .reduce((obj, key) => {
                    obj[key] = this.suffixes[key];
                    return obj;
                }, {}),
        });

    /**
     * Optional base props passed to all children routes
     * @param {{}} props
     * @returns {Route} reference to updated object
     */
    setNestedProps = props =>
        Object.assign(this, { nest: Object.assign(this.nest, { props }) });

    /**
     * Optional extra properties passed to the render of all children routes
     * @param {{}} renderProps
     * @returns {Route} reference to updated object
     */
    setNestedRenderProps = renderProps =>
        Object.assign(this, {
            nest: Object.assign(this.nest, { renderProps }),
        });

    /**
     * List of children routes, generates react-router/Route for each base * suffixes
     * @param {[Route]} routes
     * @returns {Route} reference to updated object
     */
    setNestedRoutes = routes =>
        Object.assign(this, { nest: Object.assign(this.nest, { routes }) });

    /**
     * Add single or multiple routes to the existing list of nested routes
     * @param {...Route} routes
     * @returns {Route} reference to updated object
     */
    addNestedRoutes = (...routes) =>
        Object.assign(this, {
            nest: Object.assign(this.nest, {
                routes: [...this.nest.routes, ...routes],
            }),
        });

    /**
     * Remove from list of routes only works if nested routes were named
     * @param {...string} names
     * @returns {Route} reference to updated object
     */
    removeNestedRoutes = (...names) =>
        Object.assign(this, {
            nest: Object.assign(this.nest, {
                routes: this.nest.routes.filter(
                    route => !names.includes(route.name),
                ),
            }),
        });

    _canRender = ({ path, children, component, render }) =>
        path && (children || component || render);

    _trim = ({ name, props, renderProps }) => ({
        name,
        props,
        renderProps,
    });

    _flatten = (route, base) => {
        const routes = [];
        const pathSuffixes = Object.values(route.suffixes);
        const paths = pathSuffixes.length
            ? pathSuffixes.map(suffix => `${route.props.path || ""}${suffix}`)
            : [route.props.path];
        const {
            props: nestedProps,
            renderProps: nestedRenderProps,
            routes: nestedRoutes,
        } = route.nest;
        paths.forEach(path => {
            const fullPath = `${base}${path || ""}`;
            const props = { ...route.props, path };
            if (this._canRender(props)) {
                props.path = fullPath;
                routes.push(this._trim({ ...route, props }));
            }
            nestedRoutes.forEach(nestedRoute => {
                routes.push(
                    ...this._flatten(
                        {
                            ...nestedRoute,
                            props: {
                                ...nestedProps,
                                ...nestedRoute.props,
                            },
                            renderProps: {
                                ...nestedRenderProps,
                                ...nestedRoute.renderProps,
                            },
                        },
                        fullPath,
                    ),
                );
            });
        });
        return routes;
    };

    /**
     * Build and return all the react-router/Route components
     * @param {string} base to append all routes to
     * @returns {[ReactRoute]}
     */
    render = (base = "") => {
        const routes = this._flatten(this, base);
        return routes.map(route => {
            const rProps = route.props;
            if (Object.values(route.renderProps).length) {
                rProps.render = props =>
                    rProps.component ? (
                        <rProps.component {...route.renderProps} {...props} />
                    ) : (
                        route.props.render({
                            ...route.renderProps,
                            ...props,
                        })
                    );
                delete rProps.component;
                delete rProps.children;
            }
            return <ReactRoute {...rProps} key={rProps.key || rProps.path} />;
        });
    };

    _describe = (route, base) => {
        const basePath = `${base || ""}${route.props.path || ""}`;
        const suffixNames = Object.keys(route.suffixes);
        const nextDirections = suffixNames.length
            ? suffixNames.reduce((directions, name) => {
                  const path = `${basePath}${route.suffixes[name]}`;
                  const nextRoute = {
                      ...route,
                      name,
                      props: {},
                      suffixes: {},
                  };
                  return Object.assign(directions, {
                      [name]: this._describe(nextRoute, path),
                  });
              }, {})
            : route.nest.routes.reduce((directions, nextRoute) => {
                  const nested = this._describe(nextRoute, basePath);
                  return Object.assign(
                      directions,
                      nextRoute.name ? { [nextRoute.name]: nested } : nested,
                  );
              }, {});
        return Object.assign(nextDirections, { $: basePath });
    };

    /**
     * Generate easily accessible directions to all named paths
     * @param {string} base to append all route paths to
     * @returns {{}} object with named paths
     * @example
```js
{
    $: "/mybase",
    {
        demo: {
            $: "/mybase/demo/:id",
            { aChild: { $: "/mybase/demo/:id/iam/a/child" } }
        }
    },
    {
        example: {
            $: "/mybase/example/:id",
            { aChild: { $: "/mybase/example/:id/iam/a/child" } }
        }
    },
}
```
     */
    describe = (base = "") => this._describe(this, base);
}
