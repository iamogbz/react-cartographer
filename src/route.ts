import * as React from "react";
import { Route as ReactRoute } from "react-router";
import {
    AnyObject,
    ReactRouteProps,
    RouteDirection,
    RouteShape,
    RouteRenderProps,
    Suffixes,
} from "global";

export class Route {
    _name: string;
    props: ReactRouteProps;
    renderProps: AnyObject;
    suffixes: Suffixes;
    nest: { props: ReactRouteProps; renderProps: AnyObject; routes: Route[] };
    /**
     * Create a new Route object
     * @param {{}} route shape
     */
    constructor({ name, props, renderProps, suffixes, nest }: RouteShape = {}) {
        this.name = name ? name.toString() : undefined;
        this.props = { ...props };
        this.renderProps = { ...renderProps };
        this.suffixes = { ...suffixes };
        this.nest = nest
            ? {
                  props: { ...nest.props },
                  renderProps: { ...nest.renderProps },
                  routes: (nest.routes || []).map(shape => new Route(shape)),
              }
            : { props: {}, renderProps: {}, routes: [] };
    }

    /**
     * Get name of the route
     */
    get name(): string {
        return this._name;
    }

    /**
     * Sets name of the route. Non string values are rejected and warned against.
     */
    set name(value: string) {
        this._name = value;
    }

    /**
     * Sets name of the route
     * @param {string} name
     * @returns {Route} reference to updated object
     */
    setName = (name: string): Route => Object.assign(this, { name });

    /**
     * Sets react router route props
     * @param {{}} props
     * @returns {Route} reference to updated object
     */
    setProps = (props: {}): Route => Object.assign(this, { props });

    /**
     * Adds to react router route props
     * @param {{}} props
     * @returns {Route} reference to updated object
     */
    addProps = (props: {}): Route => {
        Object.assign(this.props, props);
        return this;
    };

    rPath = (path: string) => this.addProps({ path });

    rKey = (key: string) => this.addProps({ key });

    rExact = (exact: boolean) => this.addProps({ exact });

    rStrict = (strict: boolean) => this.addProps({ strict });

    rLocation = (location: string) => this.addProps({ location });

    rSensitive = (sensitive: boolean) => this.addProps({ sensitive });

    rChildren = (children: any) => this.addProps({ children });

    rComponent = (component: React.ReactElement) =>
        this.addProps({ component });

    rRender = (render: Function) => this.addProps({ render });

    /**
     * Extra properties passed to the render of this route.
     * When this is non empty react router render property is always used.
     * Returns reference to updated object.
     */
    setRenderProps = (renderProps: AnyObject): Route =>
        Object.assign(this, { renderProps });

    /**
     * List of paths appended to base, result used as Route.path prop.
     * Returns reference to updated object.
     */
    setSuffixes = (suffixes: Suffixes): Route =>
        Object.assign(this, { suffixes });

    /**
     * Add one or more suffixes to existing route.suffixes.
     * Returns reference to updated object.
     */
    addSuffixes = (suffixes: Suffixes): Route =>
        Object.assign(this, {
            suffixes: Object.assign(this.suffixes, suffixes),
        });

    /**
     * Deletes specified suffixes using names as keys
     */
    removeSuffixes = (...names: string[]): Route =>
        Object.assign(this, {
            suffixes: Object.keys(this.suffixes)
                .filter(key => !names.includes(key))
                .reduce((obj: AnyObject, key) => {
                    obj[key] = this.suffixes[key];
                    return obj;
                }, {}),
        });

    /**
     * Optional base props passed to all children routes.
     * Returns reference to updated object.
     */
    setNestedProps = (props: AnyObject): Route =>
        Object.assign(this, { nest: Object.assign(this.nest, { props }) });

    /**
     * Optional extra properties passed to the render of all children routes.
     * Returns reference to updated object.
     */
    setNestedRenderProps = (renderProps: AnyObject): Route =>
        Object.assign(this, {
            nest: Object.assign(this.nest, { renderProps }),
        });

    /**
     * List of children routes which will generate react-router/Route
     * for every combination of base and suffixes.
     * Returns reference to updated object.
     */
    setNestedRoutes = (routes: Route[]): Route =>
        Object.assign(this, { nest: Object.assign(this.nest, { routes }) });

    /**
     * Add single or multiple routes to the existing list of nested routes.
     * Returns reference to updated object.
     */
    addNestedRoutes = (...routes: Route[]): Route =>
        Object.assign(this, {
            nest: Object.assign(this.nest, {
                routes: [...this.nest.routes, ...routes],
            }),
        });

    /**
     * Remove from list of routes only works if nested routes were named.
     * Returns reference to updated object.
     */
    removeNestedRoutes = (...names: string[]): Route =>
        Object.assign(this, {
            nest: Object.assign(this.nest, {
                routes: this.nest.routes.filter(
                    route => !names.includes(route.name),
                ),
            }),
        });

    /**
     * Check if route can be rendered using props
     */
    _canRender = ({ path, children, component, render }: ReactRouteProps): boolean =>
        Boolean(path && (children || component || render));

    /**
     * Get only the relevant props for rendering react route.
     */
    _trim = ({ name, props, renderProps }: RouteShape): RouteRenderProps => ({
        name,
        props,
        renderProps,
    });

    /**
     * Convert nested routes to list of react route renderable objects.
     */
    _flatten = (route: Route, base: string): RouteRenderProps[] => {
        const routes: RouteRenderProps[] = [];
        const pathSuffixes = Object.values(route.suffixes);
        const paths = pathSuffixes.length
            ? pathSuffixes.map(suffix => `${route.props.path || ""}${suffix}`)
            : [route.props.path];
        const {
            props: nestedProps,
            renderProps: nestedRenderProps,
            routes: nestedRoutes,
        } = route.nest;
        paths.forEach(path => {
            const fullPath = `${base}${path || ""}`;
            const props = { ...route.props, path };
            if (this._canRender(props)) {
                props.path = fullPath;
                routes.push(this._trim({ ...route, props }));
            }
            nestedRoutes.forEach(nestedRoute => {
                routes.push(
                    ...this._flatten(
                        {
                            ...nestedRoute,
                            props: {
                                ...nestedProps,
                                ...nestedRoute.props,
                            },
                            renderProps: {
                                ...nestedRenderProps,
                                ...nestedRoute.renderProps,
                            },
                        } as Route,
                        fullPath,
                    ),
                );
            });
        });
        return routes;
    };

    /**
     * Build and return all the react-router/Route components.
     */
    render = (base: string = ""): React.ReactElement[] => {
        const routes = this._flatten(this, base);
        return routes.map(route => {
            const rProps = { ...route.props };
            const ReactComponent = route.props.component;
            const renderFn: any = route.props.render || route.props.children;
            if (Object.values(route.renderProps).length) {
                rProps.render = (props: AnyObject) => {
                    const renderProps = { ...route.renderProps, ...props };
                    return ReactComponent
                        ? React.createElement(ReactComponent, renderProps)
                        : renderFn(renderProps);
                };
                delete rProps.component;
            }
            if (rProps.render) delete rProps.children;
            if (rProps.component) delete rProps.render;
            return React.createElement(ReactRoute, {
                ...rProps,
                key: String(rProps.key || rProps.path),
            });
        });
    };

    _describe = (route: Route, base: string): RouteDirection => {
        const empty: RouteDirection = { $: null };
        const basePath = `${base || ""}${route.props.path || ""}`;
        const suffixNames = Object.keys(route.suffixes);
        const nextDirections: RouteDirection = suffixNames.length
            ? suffixNames.reduce((directions, name) => {
                  const path = `${basePath}${route.suffixes[name]}`;
                  const nextRoute = {
                      ...route,
                      name,
                      props: {},
                      suffixes: {},
                  };
                  return Object.assign(directions, {
                      [name]: this._describe(nextRoute, path),
                  });
              }, empty)
            : route.nest.routes.reduce((directions, nextRoute) => {
                  const nested = this._describe(nextRoute, basePath);
                  return Object.assign(
                      directions,
                      nextRoute.name ? { [nextRoute.name]: nested } : nested,
                  );
              }, empty);
        return Object.assign(nextDirections, { $: basePath });
    };

    /**
     * Generate easily accessible directions to all named paths
     * @example
```js
{
    $: "/mybase",
    {
        demo: {
            $: "/mybase/demo/:id",
            { aChild: { $: "/mybase/demo/:id/iam/a/child" } }
        }
    },
    {
        example: {
            $: "/mybase/example/:id",
            { aChild: { $: "/mybase/example/:id/iam/a/child" } }
        }
    },
}
```
     */
    describe = (base: string = ""): RouteDirection =>
        this._describe(this, base);
}
