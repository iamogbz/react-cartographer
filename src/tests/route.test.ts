import { shallow } from "enzyme";
import { AnyObject, RouteShape } from "global";
import { Route } from "route";
import * as mocks from "./mocks";

const NOOP = () => {};

describe("Route", () => {
    let route: Route & AnyObject;

    beforeEach(() => {
        route = new Route(mocks.routeShape);
    });

    describe("initialization", () => {
        it("should use string value of name", () => {
            const mockName = "mockName";
            const name = { toString: () => mockName };
            route = new Route({ name } as RouteShape);
            expect(route.name).toEqual(mockName);
        });
        it("should initialize to empty nest", () => {
            route = new Route();
            expect(route.nest).toEqual({
                props: {},
                renderProps: {},
                routes: [],
            });
        });
    });

    describe("setName", () => {
        it("should set name and return route", () => {
            const mockName = "mock-name";
            expect(route.setName(mockName)).toBe(route);
            expect(route.name).toEqual(mockName);
        });
    });

    describe("setProps", () => {
        it("should set props and return route", () => {
            const mockProps = mocks.routeShape.nest.props;
            expect(route.setProps(mockProps)).toBe(route);
            expect(route.props).toEqual(mockProps);
        });

        it("should add props and return route", () => {
            const mockProps = mocks.routeShape.nest.props;
            expect(route.addProps(mockProps)).toBe(route);
            expect(route.props).toMatchObject(mockProps);
        });

        it("should call addProps", () => {
            const addPropsSpy = jest.spyOn(route, "addProps");
            Object.entries({
                children: NOOP,
                component: NOOP,
                exact: false,
                key: "mock-key",
                location: { pathname: "/mock/pathname" },
                path: "/mock/path",
                render: NOOP,
                sensitive: true,
                strict: true,
            }).forEach(([key, value]) => {
                const method = `r${key[0].toUpperCase()}${key.substring(1)}`;
                route[method](value);
                expect(addPropsSpy).toBeCalledWith({ [key]: value });
                addPropsSpy.mockReset();
            });
        });
    });

    describe("setRenderProps", () => {
        it("should assign renderProps and return route", () => {
            const mockRenderProps = { mockKey: "mockValue" };
            expect(route.setRenderProps(mockRenderProps)).toBe(route);
            expect(route.renderProps).toEqual(mockRenderProps);
        });
    });

    describe("setSuffixes", () => {
        it("should assign suffixes and return route", () => {
            const mockSuffixes = { mockName: "mockPath" };
            expect(route.setSuffixes(mockSuffixes)).toBe(route);
            expect(route.suffixes).toEqual(mockSuffixes);
        });

        it("should append suffixes and return route", () => {
            const mockSuffixes = { mockName: "/mockPath" };
            expect(route.addSuffixes(mockSuffixes)).toBe(route);
            expect(route.suffixes).toEqual(mockSuffixes);
            const extraSuffixes = { extraKey: "/another/path" };
            route.addSuffixes(extraSuffixes);
            expect(route.suffixes).toEqual({
                ...mockSuffixes,
                ...extraSuffixes,
            });
        });

        it("should remove suffixes and return route", () => {
            const mockSuffixes = {
                mockName: "/mockPath",
                extraKey: "/another/path",
            };
            route.setSuffixes(mockSuffixes);
            expect(route.suffixes).toEqual(mockSuffixes);
            expect(route.removeSuffixes("mockName")).toBe(route);
            expect(route.suffixes).toEqual({ extraKey: "/another/path" });
        });
    });

    describe("setNestedProps", () => {
        it("should assign nest props and return route", () => {
            const mockNestProps = mocks.routeShape.props;
            expect(route.setNestedProps(mockNestProps)).toBe(route);
            expect(route.nest.props).toEqual(mockNestProps);
        });
    });

    describe("setNestedRenderProps", () => {
        it("should assign nest renderProps and return route", () => {
            const mockNestRenderProps = mocks.routeShape.renderProps;
            expect(route.setNestedRenderProps(mockNestRenderProps)).toBe(route);
            expect(route.nest.renderProps).toEqual(mockNestRenderProps);
        });
    });

    describe("setNestedRoutes", () => {
        it("should assign nest routes and return route", () => {
            route.nest.routes = [];
            const mockNestedRoutes = mocks.routeShape.nest.routes;
            const mockRoutes = mockNestedRoutes.map(shape => new Route(shape));
            expect(route.setNestedRoutes(mockRoutes)).toBe(route);
            expect(route.nest.routes).toEqual(mockRoutes);
        });

        it("should append nest routes and return route", () => {
            const extraRoute = new Route({ props: { path: "/mockPath" } });
            expect(route.addNestedRoutes(extraRoute)).toBe(route);
            expect(route.nest.routes).toContain(extraRoute);
        });

        it("should remove nest routes and return route", () => {
            const extraRoute = new Route({ props: { path: "/mockPath" } });
            route.addNestedRoutes(extraRoute);
            const numNestedRoutes = route.nest.routes.length;
            expect(route.removeNestedRoutes("twin", "son")).toBe(route);
            expect(route.nest.routes).toContain(extraRoute);
            expect(route.nest.routes.length).toBe(numNestedRoutes - 2);
        });
    });

    describe("render", () => {
        const mockProps = {
            mockProp: "mock-value",
            path: "/mock-path",
        };
        const mockRenderProps = { mockRenderProp: "mock-render-value" };
        const newMockRoute = (props = {}, renderProps = {}) =>
            new Route({
                props: { ...mockProps, ...props },
                renderProps,
            });

        it("should match snapshot", () => {
            expect(route.render()).toMatchSnapshot();
        });

        it("should use react route render with component when render props set", () => {
            const component = jest.fn(NOOP);
            const mockRoute = newMockRoute({ component }, mockRenderProps);
            const reactRoute = mockRoute.render().pop();
            expect(reactRoute.props.component).toBeUndefined();
            shallow(reactRoute.props.render());
            expect(component).toBeCalledWith(mockRenderProps, {});
        });

        it("should use render with render fn when render props set", () => {
            const render = jest.fn();
            const mockRoute = newMockRoute({ render }, mockRenderProps);
            const reactRoute = mockRoute.render().pop();
            expect(reactRoute.props.component).toBeUndefined();
            reactRoute.props.render();
            expect(render).toBeCalledWith(mockRenderProps);
        });

        it("should render nothing when no render fn", () => {
            const mockRoute = newMockRoute({}, mockRenderProps);
            expect(mockRoute.render().pop()).toBeUndefined();
        });

        it("should use component over render and children", () => {
            const component = jest.fn(NOOP);
            const render = jest.fn();
            const children = jest.fn();
            const mockRoute = newMockRoute({
                component,
                render,
                children,
            });
            const reactRoute = mockRoute.render().pop();
            expect(reactRoute).toMatchSnapshot();
        });

        it("should use render over children", () => {
            const render = jest.fn();
            const children = jest.fn();
            const mockRoute = newMockRoute({
                render,
                children,
            });
            const reactRoute = mockRoute.render().pop();
            expect(reactRoute).toMatchSnapshot();
        });

        it("should use children if nothing else", () => {
            const children = jest.fn();
            const mockRoute = newMockRoute({ children });
            const reactRoute = mockRoute.render().pop();
            expect(reactRoute).toMatchSnapshot();
        });
    });

    describe("describe", () =>
        it("should match snapshot", () =>
            expect(route.describe()).toMatchSnapshot()));
});
