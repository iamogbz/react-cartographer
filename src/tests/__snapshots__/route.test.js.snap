// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route should describe routes correctly 1`] = `
Object {
  "$": "",
  "son": Object {
    "$": "/child/son",
  },
  "twin": Object {
    "$": "/twin",
    "first": Object {
      "$": "/twin-one",
    },
    "second": Object {
      "$": "/twin-two",
    },
  },
}
`;

exports[`Route should render routes correctly 1`] = `
Array [
  <Route
    exact={true}
    path="/twin-one"
    render={[Function]}
  />,
  <Route
    exact={true}
    path="/twin-two"
    render={[Function]}
  />,
  <Route
    exact={true}
    path="/child/son"
    render={[MockFunction]}
  />,
]
`;
